src/osemgrep/language_server/scan_helpers/Scan_helpers.ml

```diff
@@ -64,56 +60,66 @@ let run_semgrep ?(targets = None) ?(rules = None) ?(git_ref = None)
+        results files
+    in
+    end_progress server token;
+    batch_notify server diagnostics
   in
-  end_progress server token;
-  batch_notify server diagnostics
+  wrap_with_detach f
 
 (** Scan a single file. Passing [content] will write it to a temp file,
    and scan that temp file, then return results as if [uri] was scanned *)
 let scan_file ?(content = None) server uri =
-  let file_path = Uri.to_path uri in
-  let file = Fpath.v file_path in
-  let targets, git_ref =
-    match content with
-    | None -> ([ file ], None)
-    | Some content ->
-        let name = Fpath.basename file in
-        let ext = Fpath.get_ext file in
-        let tmp_file = Common.new_temp_file name ext in
-        Common.write_file tmp_file content;
-        ([ Fpath.v tmp_file ], Some Fpath.(to_string file))
-  in
-  let session_targets = Session.targets server.session in
-  let targets = if List.mem file session_targets then targets else [] in
-  let targets = Some targets in
-  let results, _ = run_semgrep ~git_ref ~targets server in
-  let results =
-    match content with
-    | Some _ ->
-        let existing_results, _ = run_semgrep ~targets:(Some [ file ]) server in
-        results @ existing_results
-    | None -> results
-  in
-  let results =
-    Common.map (fun (m : Out.cli_match) -> { m with path = file_path }) results
-  in
-  let files = [ file ] in
-  Session.record_results server.session results files;
-  let diagnostics =
-    Diagnostics.diagnostics_of_results ~is_intellij:server.session.is_intellij
-      results files
+  let f () =
+    let file_path = Uri.to_path uri in
+    let file = Fpath.v file_path in
+    let targets, git_ref =
+      match content with
+      | None -> ([ file ], None)
+      | Some content ->
+          let name = Fpath.basename file in
+          let ext = Fpath.get_ext file in
+          let tmp_file = Common.new_temp_file name ext in
+          Common.write_file tmp_file content;
+          ([ Fpath.v tmp_file ], Some Fpath.(to_string file))
+    in
+    let session_targets = Session.targets server.session in
+    let targets = if List.mem file session_targets then targets else [] in
+    let targets = Some targets in
+    let results, _ = run_semgrep ~git_ref ~targets server in
+    let results =
+      match content with
+      | Some _ ->
+          let existing_results, _ =
+            run_semgrep ~targets:(Some [ file ]) server
+          in
+          results @ existing_results
+      | None -> results
+    in
+    let results =
+      Common.map
+        (fun (m : Out.cli_match) -> { m with path = file_path })
+        results
+    in
+    let files = [ file ] in
+    Session.record_results server.session results files;
+    let diagnostics =
+      Diagnostics.diagnostics_of_results ~is_intellij:server.session.is_intellij
+        results files
+    in
+    batch_notify server diagnostics
   in
-  batch_notify server diagnostics
+  wrap_with_detach f
```

Comment:
same here, small comment explaining what could take a long time in f() and block the whole lwt event loop.